storage/index.go

```diff
@@ -207,15 +210,28 @@ func (i *nsIndex) Query(
 		return index.QueryResults{}, errDbIndexUnableToQueryClosed
 	}
 
+	reader, err := i.active.segment.Reader()
+	if err != nil {
+		return index.QueryResults{}, err
+	}
+
+	readers := []m3ninxindex.Reader{reader}
+	exec := executor.NewExecutor(readers)
+
 	// FOLLOWUP(prateek): push down QueryOptions to restrict results
-	iter, err := i.active.segment.Query(query.Query)
+	iter, err := exec.Execute(query.Query.SearchQuery())
 	i.RUnlock()
 	if err != nil {
+		exec.Close()
+		reader.Close()
 		return index.QueryResults{}, err
 	}
 
 	return index.QueryResults{
-		Iterator:   index.NewIterator(i.nsID, iter, i.opts),
+		Iterator: index.NewIterator(i.nsID, iter, i.opts, func() {
+			exec.Close()
+			reader.Close()
```

Comment:
We can update this, and it wouldn't actually matter because we don't check the error here, but the `executor` actually takes ownership of the `readers` so it will be closed by the `exectutor`.

Reply:
cheers, will do